extensions/pkg/controller/worker/cloudprofile.go

```diff
@@ -0,0 +1,120 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package worker
+
+import (
+	"fmt"
+	"slices"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+)
+
+// CapabilitiesAccessor defines an interface for retrieving Capabilities.
+type CapabilitiesAccessor interface {
+	GetCapabilities() v1beta1.Capabilities
+}
+
+// FindBestImageFlavor finds the most appropriate image version flavor based on the requested machine capabilities.
+// The provided capability definitions are used for applying defaults.
+func FindBestImageFlavor[T CapabilitiesAccessor](
+	providerImageFlavors []T,
+	machineCapabilities v1beta1.Capabilities,
+	capabilityDefinitions []v1beta1.CapabilityDefinition,
+) (T, error) {
+	var zeroValue T
+
+	compatibleFlavors := filterCompatibleImageFlavors(providerImageFlavors, machineCapabilities, capabilityDefinitions)
+	if len(compatibleFlavors) == 0 {
+		return zeroValue, fmt.Errorf("no compatible flavor found")
+	}
+
+	bestFlavor, err := selectBestImageFlavor(compatibleFlavors, capabilityDefinitions)
+	if err != nil {
+		return zeroValue, err
+	}
+	return bestFlavor, nil
+}
+
+// filterCompatibleImageFlavors returns all image capabilityFlavors that are compatible with the given machine capabilities.
+func filterCompatibleImageFlavors[T CapabilitiesAccessor](
+	imageFlavors []T,
+	machineCapabilities v1beta1.Capabilities,
+	capabilityDefinitions []v1beta1.CapabilityDefinition,
+) []T {
+	var compatibleFlavors []T
+	for _, imageFlavor := range imageFlavors {
+		if v1beta1helper.AreCapabilitiesCompatible(imageFlavor.GetCapabilities(), machineCapabilities, capabilityDefinitions) {
+			compatibleFlavors = append(compatibleFlavors, imageFlavor)
+		}
+	}
+	return compatibleFlavors
+}
+
+// selectBestImageFlavor selects the most appropriate image flavor based on the priority
+// of its capabilities and their values as defined in capabilityDefinitions.
+//
+// Selection follows a priority-based approach:
+// 1. Capabilities are ordered by priority in the definitions list (highest priority first)
+// 2. Within each capability, values are ordered by preference (most preferred first)
+// 3. Selection is determined by the first capability value difference found
+func selectBestImageFlavor[T CapabilitiesAccessor](
+	compatibleSets []T,
+	capabilityDefinitions []v1beta1.CapabilityDefinition,
+) (T, error) {
+	var zeroValue T
```

Comment:
I guess for improved readability we could also move this down to `:116` to its only usage.